File: DrawDemoSwift/Old ObjC/GP for GCDrawDemo/GCSGradientCell.m
//
//  GCGradientCell.m
//  GradientTest
//
//  Created by Jason Jobe on 3/3/07.
//  Copyright 2007 __MyCompanyName__. All rights reserved.
//

#import "GCSGradientCell.h"

#import "GCSGradientWell.h"
#import "GCSMiniCircularSlider.h"
#import "GCSMiniControlCluster.h"
#import "GCSMiniRadialControl2.h"
#import "GCSMiniRadialControls.h"

#import <DKDrawKit/DKGradientExtensions.h>
#import <DKDrawKit/LogEvent.h>

#pragma mark Contants (Non-localized)
// private IDs locate mini controls within the cell

static NSString *kLinearAngleControlID = @"kLinearAngleControlID";
static NSString *kRadialStartControlID = @"kRadialStartControlID";
static NSString *kRadialEndControlID = @"kRadialEndControlID";
static NSString *kSweepCentreControlID = @"kSweepCentreControlID";
static NSString *kSweepSegmentsControlID = @"kSweepSegmentsControlID";
static NSString *kSweepAngleControlID = @"kSweepAngleControlID";

// clusters:

static NSString *kLinearControlsClusterID = @"kLinearControlsClusterID";
static NSString *kRadialControlsClusterID = @"kRadialControlsClusterID";
static NSString *kSweepControlsClusterID = @"kSweepControlsClusterID";

#pragma mark Static Vars
static NSInteger sMFlags = 0;

@implementation GCSGradientCell
#pragma mark As a GCGradientCell
- (void)setupMiniControls
{
	// the mini controls are stored in a series of hierarchical clusters. The top level is basically just used as a
	// container for the lower level clusters. Each subcluster contains a group of mini controls, one for each of the
	// gradient modes/types

	mMiniControls = [[GCSMiniControlCluster alloc] initWithBounds:NSZeroRect inCluster:nil];
	mMiniControls.delegate = self;
	[mMiniControls forceVisible:NO];

	GCSMiniControlCluster *mcc;
	GCSMiniControl *mini;

	// first contains circular slider for linear gradient angle

	mcc = [[GCSMiniControlCluster alloc] initWithBounds:NSZeroRect inCluster:mMiniControls];
	[mcc setIdentifier:kLinearControlsClusterID];

	mini = [[GCSMiniCircularSlider alloc] initWithBounds:NSZeroRect inCluster:mcc];
	[mini setIdentifier:kLinearAngleControlID];

	// second has twin radial controls

	mcc = [[GCSMiniControlCluster alloc] initWithBounds:NSZeroRect inCluster:mMiniControls];
	[mcc setIdentifier:kRadialControlsClusterID];

	// allow shift key to move both minicontrols together:

	[mcc setLinkControlPart:kDKRadial2HitIris modifierKeyMask:NSEventModifierFlagShift];

	mini = [[GCSMiniRadialControl2 alloc] initWithBounds:NSZeroRect inCluster:mcc];
	[mini setIdentifier:kRadialEndControlID];

	mini = [[GCSMiniRadialControl2 alloc] initWithBounds:NSZeroRect inCluster:mcc];
	[mini setIdentifier:kRadialStartControlID];

	// third has circular slider + single radial control + straight slider
	// n.b. order is important as controls overlap. hit testing is done in reverse order to
	// that here, which is the drawing order.

	mcc = [[GCSMiniControlCluster alloc] initWithBounds:NSZeroRect inCluster:mMiniControls];
	[mcc setIdentifier:kSweepControlsClusterID];

	mini = [[GCSMiniCircularSlider alloc] initWithBounds:NSZeroRect inCluster:mcc];
	[mini setIdentifier:kSweepAngleControlID];

	mini = [[GCSMiniSlider alloc] initWithBounds:NSZeroRect inCluster:mcc];
	[mini setIdentifier:kSweepSegmentsControlID];
	[mini setInfoWindowMode:kDKMiniControlInfoWindowCentred];
	[mini setInfoWindowFormat:@"0"];

	mini = [[GCSMiniRadialControls alloc] initWithBounds:NSZeroRect inCluster:mcc];
	[mini setIdentifier:kSweepCentreControlID];
}

- (void)setControlledAttributeFromMiniControl:(GCSMiniControl *)ctrl
{
	NSString *ident = [ctrl identifier];

	GCSMiniRadialControls *rc;

	if ([ident isEqualToString:kLinearAngleControlID] || [ident isEqualToString:kSweepAngleControlID]) {
		[self gradient].angle = ctrl.value;
	} else if ([ident isEqualToString:kSweepSegmentsControlID]) {
		NSInteger seg = ctrl.value * 50;
		if (seg < 4)
			seg = 0;

		//[[self gradient] setNumberOfAngularSegments:seg];
	} else if ([ident isEqualToString:kRadialStartControlID] || [ident isEqualToString:kSweepCentreControlID]) {
		rc = (GCSMiniRadialControls *)ctrl;

		//	LogEvent_(kStateEvent, @"setting starting radius: %f", [rc radius]);

		NSPoint p = [[self gradient] mapPoint:rc.centre fromRect:mControlBoundsRect];
		[[self gradient] setRadialStartingPoint:p];
		[[self gradient] setRadialStartingRadius:rc.radius / mControlBoundsRect.size.width];
	} else if ([ident isEqualToString:kRadialEndControlID]) {
		rc = (GCSMiniRadialControls *)ctrl;

		//	LogEvent_(kStateEvent, @"setting ending radius: %f", [rc radius]);
		NSPoint p = [[self gradient] mapPoint:rc.centre fromRect:mControlBoundsRect];
		[[self gradient] setRadialEndingPoint:p];
		[[self gradient] setRadialEndingRadius:rc.radius / mControlBoundsRect.size.width];
	}
}

#pragma mark -
- (void)setMiniControlBoundsWithCellFrame:(NSRect)cellframe forMode:(DKSGradientWellMode)mode
{
	// sets up the mini controls' bounds from the cellFrame. Each one is individually calculated as appropriate. Note
	// that some types, notably the circular slider, position themselves centrally in their bounds so this method need
	// not bother with that.

	NSRect cframe = NSInsetRect(cellframe, 20, 20);
	mMiniControls.view = self.controlView;

	// linear:

	if (mode == kDKGradientWellAngleMode) {
		[self setMiniControlBounds:cframe withIdentifier:kLinearAngleControlID];
	} else if (mode == kDKGradientWellRadialMode) {
		// radial controls likewise just need the entire frame:

		[self setMiniControlBounds:cellframe withIdentifier:kRadialStartControlID];
		[self setMiniControlBounds:cellframe withIdentifier:kRadialEndControlID];
	} else if (mode == kDKGradientWellSweepMode) {
		// sweep controls:

		[self setMiniControlBounds:cframe withIdentifier:kSweepAngleControlID];
		[self setMiniControlBounds:cellframe withIdentifier:kSweepCentreControlID];

		// only the segment slider has a complex bounds calc:

		NSRect sr = NSInsetRect(cframe, 40, 50);
		sr.size.height = 12;
		sr.origin.y = cframe.origin.y + (cframe.size.height * 0.63);

		[self setMiniControlBounds:sr withIdentifier:kSweepSegmentsControlID];
	}
}

- (void)setMiniControlBounds:(NSRect)br withIdentifier:(NSString *)key
{
	// sets the mini control with <key> identifier to have the bounds <br>

	GCSMiniControl *mc = [mMiniControls controlForKey:key];

	if (mc)
		mc.bounds = br;
}

- (void)drawMiniControlsForMode:(DKSGradientWellMode)mode
{
	// given the mode which is set by the owning GCGradientWell, this draws the controls in the appropriate cluster.

	[[self controlClusterForMode:mode] draw];
}

- (GCSMiniControlCluster *)controlClusterForMode:(DKSGradientWellMode)mode
{
	switch (mode) {
		default:
		case kDKGradientWellDisplayMode:
			return nil;

		case kDKGradientWellAngleMode:
			return (GCSMiniControlCluster *)[mMiniControls controlForKey:kLinearControlsClusterID];

		case kDKGradientWellRadialMode:
			return (GCSMiniControlCluster *)[mMiniControls controlForKey:kRadialControlsClusterID];

		case kDKGradientWellSweepMode:
			return (GCSMiniControlCluster *)[mMiniControls controlForKey:kSweepControlsClusterID];
	}
}

- (GCSMiniControl *)miniControlForIdentifier:(NSString *)key
{
	return [mMiniControls controlForKey:key];
}

- (void)updateMiniControlsForMode:(DKSGradientWellMode)mode
{
	// sets the minicontrol values in <mode> cluster to match the current gradient

	mUpdatingControls = YES;

	switch (mode) {
		case kDKGradientWellAngleMode:
			[self miniControlForIdentifier:kLinearAngleControlID].value = [self gradient].angle;
			break;

		case kDKGradientWellRadialMode:
			//	LogEvent_(kStateEvent, @"setting up radial controls");
			{
				GCSMiniRadialControl2 *rc = (GCSMiniRadialControl2 *)[mMiniControls controlForKey:kRadialStartControlID];

				[rc setRingRadiusScale:0.85];
				rc.centre = [[self gradient] mapPoint:[[self gradient] radialStartingPoint] toRect:mControlBoundsRect];
				rc.radius = [[self gradient] radialStartingRadius] * mControlBoundsRect.size.width;
				rc.tabColor = [[self gradient] colorAtValue:0.0];

				rc = (GCSMiniRadialControl2 *)[mMiniControls controlForKey:kRadialEndControlID];

				rc.centre = [[self gradient] mapPoint:[[self gradient] radialEndingPoint] toRect:mControlBoundsRect];
				rc.radius = [[self gradient] radialEndingRadius] * mControlBoundsRect.size.width;
				rc.tabColor = [[self gradient] colorAtValue:1.0];
			}
			break;

		case kDKGradientWellSweepMode: {
			// sweep controls:

			GCSMiniRadialControls *rc = (GCSMiniRadialControls *)[mMiniControls controlForKey:kSweepCentreControlID];

			rc.centre = [[self gradient] mapPoint:[[self gradient] radialStartingPoint] toRect:mControlBoundsRect];

			NSInteger seg = 100; //[[self gradient] numberOfAngularSegments];
			CGFloat v = (CGFloat)seg / 50.0;

			if (seg < 4)
				v = 0.0;

			[self miniControlForIdentifier:kSweepSegmentsControlID].value = v;
			[self miniControlForIdentifier:kSweepAngleControlID].value = [self gradient].angle;
		} break;

		default:
			break;
	}

	mUpdatingControls = NO;
}

#pragma mark -
- (NSRect)proxyIconRectInCellFrame:(NSRect)rect
{
	NSImage *ficon = [NSImage imageNamed:@"fileiconsmall"];
	NSRect ir, br;

	br = NSInsetRect(rect, 8, 9);

	ir.size = ficon.size;
	ir.origin.x = NSMaxX(br) - ir.size.width;
	ir.origin.y = NSMaxY(br) - ir.size.height;

	return ir;
}

#pragma mark -
- (void)setControlVisible:(BOOL)vis
{
	mMiniControls.visible = vis;
	[self.controlView setNeedsDisplay:YES];
}

#pragma mark -
#pragma mark As an DKGradientCell
- (SEL)action
{
	return mAction;
}

- (void)drawInteriorWithFrame:(NSRect)cellFrame inView:(NSView *)controlView
{
	if ([self gradient]) {
		[super drawInteriorWithFrame:cellFrame inView:controlView];

		mControlBoundsRect = cellFrame;
		id control = self.controlView;

		if ([control isKindOfClass:[GCSGradientWell class]]) {
			[control setupTrackingRect];
			[self setMiniControlBoundsWithCellFrame:cellFrame forMode:[control controlMode]];
			[self updateMiniControlsForMode:[control controlMode]];

			[NSBezierPath clipRect:NSInsetRect(cellFrame, 8, 8)];
			[self drawMiniControlsForMode:[control controlMode]];

			// if proxy icon flag set, draw it

			if ([control displaysProxyIcon]) {
				NSImage *ficon = [NSImage imageNamed:@"fileiconsmall"];
				[ficon setFlipped:YES];
				[ficon drawInRect:[self proxyIconRectInCellFrame:cellFrame] fromRect:NSZeroRect operation:NSCompositingOperationSourceAtop fraction:0.8];
			}
		}
	}
}

- (void)setAction:(SEL)action
{
	mAction = action;
}

- (void)setTarget:(id)target
{
	mTargetRef = target;
}

- (id)target
{
	return mTargetRef;
}

#pragma mark -
#pragma mark As an NSCell
- (BOOL)continueTracking:(NSPoint)lastPoint at:(NSPoint)currentPoint inView:(NSView *)controlView
{
#pragma unused(lastPoint)
	if ([controlView isKindOfClass:[GCSGradientWell class]]) {
		if (mHitPart == kDKHitMiniControl) {
			DKSGradientWellMode cmode = ((GCSGradientWell *)controlView).controlMode;
			GCSMiniControlCluster *cc = [self controlClusterForMode:cmode];

			return [cc mouseDraggedAt:currentPoint inPart:0 modifierFlags:self.mouseDownFlags];
		}
	}
	return NO;
}

- (NSInteger)mouseDownFlags
{
	return sMFlags;
}

- (BOOL)startTrackingAt:(NSPoint)startPoint inView:(NSView *)controlView
{
	//	LogEvent_(kReactiveEvent, @"cell starting tracking...");

	mHitPart = kDKHitOther;

	if ([controlView isKindOfClass:[GCSGradientWell class]]) {
		// hit in proxy icon?

		if (((GCSGradientWell *)controlView).displaysProxyIcon) {
			NSRect ir = [self proxyIconRectInCellFrame:mControlBoundsRect];

			if (NSPointInRect(startPoint, ir)) {
				mHitPart = kDKHitProxyIcon;
				return YES;
			}
		}

		NSInteger cmode = ((GCSGradientWell *)controlView).controlMode;
		GCSMiniControlCluster *cc = [self controlClusterForMode:cmode];

		if ([cc mouseDownAt:startPoint inPart:0 modifierFlags:self.mouseDownFlags]) {
			mHitPart = kDKHitMiniControl; // for any mini-control
			return YES;
		}
	}

	return YES;
}

- (void)stopTracking:(NSPoint)lastPoint at:(NSPoint)stopPoint inView:(NSView *)controlView mouseIsUp:(BOOL)flag
{
#pragma unused(lastPoint, flag)
	if ([controlView isKindOfClass:[GCSGradientWell class]]) {
		if (mHitPart == kDKHitMiniControl) {
			int cmode = ((GCSGradientWell *)controlView).controlMode;
			GCSMiniControlCluster *cc = [self controlClusterForMode:cmode];

			[cc mouseUpAt:stopPoint inPart:0 modifierFlags:self.mouseDownFlags];
		}
	}
}

- (BOOL)trackMouse:(NSEvent *)theEvent inRect:(NSRect)cellFrame ofView:(NSView *)controlView untilMouseUp:(BOOL)untilMouseUp
{
#pragma unused(cellFrame, untilMouseUp)
	NSPoint p = [controlView convertPoint:theEvent.locationInWindow fromView:nil];

	sMFlags = theEvent.modifierFlags;
	[mMiniControls flagsChanged:sMFlags];
	[mMiniControls setVisible:YES];

	if ([self startTrackingAt:p inView:controlView]) {
		NSEvent *event = nil;
		BOOL loop = YES;
		NSPoint currentPoint, lastPoint;

		mEnableCache = NO;
		NSEventMask mask = NSEventMaskLeftMouseUp | NSEventMaskLeftMouseDragged | NSEventMaskFlagsChanged;
		lastPoint = p;

		//	LogEvent_(kReactiveEvent, @"starting track loop, hit part = %d", mHitPart );

		while (loop) {
			event = [controlView.window nextEventMatchingMask:mask];

			//	LogEvent_(kUIEvent, @"event = %@", event);

			currentPoint = [controlView convertPoint:event.locationInWindow fromView:nil];
			sMFlags = event.modifierFlags;

			switch (event.type) {
				case NSEventTypeLeftMouseUp:
					//	LogEvent_(kReactiveEvent, @"mouse up");
					[self stopTracking:lastPoint at:currentPoint inView:controlView mouseIsUp:YES];
					loop = NO;
					break;

				case NSEventTypeLeftMouseDragged:
					loop = [self continueTracking:lastPoint at:currentPoint inView:controlView];
					if (!loop) {
						[self stopTracking:lastPoint at:currentPoint inView:controlView mouseIsUp:NO];
						if (mHitPart == kDKHitOther) {
							[mMiniControls setVisible:NO];
							[controlView initiateGradientDragWithEvent:theEvent];
						}
					}
					break;

				case NSEventTypeFlagsChanged:
					[mMiniControls flagsChanged:event.modifierFlags];
					break;

				default:
					break;
			}

			lastPoint = currentPoint;
		}
		[controlView.window discardEventsMatchingMask:mask beforeEvent:event];
		mHitPart = kDKHitNone;
		mEnableCache = YES;
		[mMiniControls flagsChanged:0];
	}
	//	LogEvent_(kReactiveEvent, @"cell ended tracking");

	return YES;
}

#pragma mark -
#pragma mark As an NSObject

- (instancetype)init
{
	self = [super init];
	if (self != nil) {
		NSAssert(NSEqualRects(mControlBoundsRect, NSZeroRect), @"Expected init to zero");
		[self setupMiniControls];
		NSAssert(!mUpdatingControls, @"Expected init to zero");
		NSAssert(mHitPart == kDKHitNone, @"Expected init to zero");

		if (mMiniControls == nil) {
			return nil;
		}
		NSAssert([self isContinuous], @"Expected isContinuous set in base class");
	}
	return self;
}

#pragma mark -
#pragma mark As a GCMiniControl delegate
- (void)miniControl:(GCSMiniControl *)mc didChangeValue:(id)newValue
{
#pragma unused(newValue)
	// delegate method called for a change in any mini-control value. route the result to the appropriate
	// setting. Note - no need to call for redisplay, that has been done.

	if (!mUpdatingControls) {
		//	LogEvent_(kInfoEvent, @"miniControl '%@' didChangeValue '%@'", [mc identifier],  newValue);

		[self setControlledAttributeFromMiniControl:mc];

		if ([self.controlView isKindOfClass:[GCSGradientWell class]])
			[(GCSGradientWell *)self.controlView syncGradientToControlSettings];
	}
}

- (CGFloat)miniControlWillUpdateInfoWindow:(GCSMiniControl *)mc withValue:(CGFloat)val
{
	if ([[mc identifier] isEqualToString:kSweepSegmentsControlID]) {
		int seg = mc.value * 50;
		if (seg < 4)
			seg = 0;

		return seg;
	} else
		return val;
}

@end

File: DrawDemoSwift/GP for DrawDemoSwift/GCSGradientCell.swift
//
//  GCSGradientCell.swift
//  DrawDemoSwift
//
//  Created by C.W. Betts on 12/19/17.
//  Copyright © 2017 DrawKit. All rights reserved.
//

import Cocoa
import DKDrawKit.DKGradientExtensions


// MARK: Contants (Non-localized)
// private IDs locate mini controls within the cell

private let kLinearAngleControlID = "kLinearAngleControlID";
private let kRadialStartControlID = "kRadialStartControlID";
private let kRadialEndControlID = "kRadialEndControlID";
private let kSweepCentreControlID = "kSweepCentreControlID";
private let kSweepSegmentsControlID = "kSweepSegmentsControlID";
private let kSweepAngleControlID = "kSweepAngleControlID";

// clusters:

private let kLinearControlsClusterID = "kLinearControlsClusterID";
private let kRadialControlsClusterID = "kRadialControlsClusterID";
private let kSweepControlsClusterID = "kSweepControlsClusterID";

// MARK: Static Vars
private var sMFlags: UInt = 0

final class GCSGradientCell: DKSGradientCell, GCMiniControlDelegate {
	/// internal "partcodes" for where a mouse hit occurred
	enum HitPart: Int {
		case none = 0
		case miniControl = 5
		case proxyIcon = 7
		case other = 999
	}
	var controlBoundsRect = NSRect.zero
	var miniControls: GCMiniControlCluster?
	var updatingControls = false
	private var hitPart = HitPart.none

	// MARK: As a GCGradientCell
	
	/// the mini controls are stored in a series of hierarchical clusters. The top level is basically just used as a
	/// container for the lower level clusters. Each subcluster contains a group of mini controls, one for each of the
	/// gradient modes/types.
	func setupMiniControls() {
		guard let mMiniControls = GCMiniControlCluster(bounds: .zero, in: nil) else {
			fatalError()
		}
		miniControls = mMiniControls
		mMiniControls.delegate = self
		mMiniControls.forceVisible(false)
	
		var mcc: GCMiniControlCluster?
		var mini: GCMiniControl?
		
		// first contains circular slider for linear gradient angle
		mcc = GCMiniControlCluster(bounds: .zero, in: mMiniControls)
		mcc?.identifier = kLinearControlsClusterID
		
		mini = GCMiniCircularSlider(bounds: .zero, in: mcc)
		mini?.identifier = kLinearAngleControlID
		
		// second has twin radial controls

		mcc = GCMiniControlCluster(bounds: .zero, in: mMiniControls)
		mcc?.identifier = kRadialControlsClusterID
		
		// allow shift key to move both minicontrols together:

		mcc?.setLinkControlPart(kDKRadial2HitIris, modifierKeyMask: .shift)
		
		mini = GCMiniRadialControl2(bounds: .zero, in: mcc)
		mini?.identifier = kRadialEndControlID
		
		mini = GCMiniRadialControl2(bounds: .zero, in: mcc)
		mini?.identifier = kRadialStartControlID

		// third has circular slider + single radial control + straight slider
		// n.b. order is important as controls overlap. hit testing is done in reverse order to
		// that here, which is the drawing order.

		mcc = GCMiniControlCluster(bounds: .zero, in: mMiniControls)
		mcc?.identifier = kSweepControlsClusterID
		
		mini = GCMiniCircularSlider(bounds: .zero, in: mcc)
		mini?.identifier = kSweepAngleControlID
		
		mini = GCMiniSlider(bounds: .zero, in: mcc)
		mini?.identifier = kSweepSegmentsControlID
		mini?.setInfoWindowMode(.miniControlInfoWindowCentred)
		mini?.setInfoWindowFormat("0")
		
		mini = GCMiniRadialControls(bounds: .zero, in: mcc)
		mini?.identifier = kSweepCentreControlID
	}
	
	func setControlledAttribute(from ctrl: GCMiniControl) {
		let ident = ctrl.identifier!
		
		switch ident {
		case kLinearAngleControlID, kSweepAngleControlID:
			gradient?.angle = ctrl.value
			
		case kSweepSegmentsControlID:
			var seg = Int(floor(ctrl.value * 50))
			
			if seg < 4 {
				seg = 0
			}
			
			//[[self gradient] setNumberOfAngularSegments:seg];

		case kRadialStartControlID, kSweepCentreControlID:
			let rc = ctrl as! GCMiniRadialControls
			
			LogEvent(.stateEvent, "setting starting radius: \(rc.radius)")

			if let grad = gradient {
				let p = grad.mapPoint(rc.centre, from: controlBoundsRect)
				grad.radialStartingPoint = p
				grad.radialStartingRadius = rc.radius / controlBoundsRect.size.width
			}
			
		case kRadialEndControlID:
			let rc = ctrl as! GCMiniRadialControls
			
			LogEvent(.stateEvent, "setting ending radius: \(rc.radius)")

			if let grad = gradient {
				let p = grad.mapPoint(rc.centre, from: controlBoundsRect)
				grad.radialEndingPoint = p
				grad.radialEndingRadius = rc.radius / controlBoundsRect.size.width
			}
			
		default:
			break
		}
	}
	
	// MARK: -
	
	/// sets up the mini controls' bounds from the cellFrame. Each one is individually calculated as appropriate. Note
	/// that some types, notably the circular slider, position themselves centrally in their bounds so this method need
	/// not bother with that.
	func setMiniControlBounds(cellFrame cellframe: NSRect, for mode: GCSGradientWell.Mode) {
		let cframe = cellframe.insetBy(dx: 20, dy: 20)
		miniControls?.view = self.controlView
		
		// linear:

		switch mode {
		case .angle:
			setMiniControlBounds(cframe, identifier: kLinearAngleControlID)
			
		case .radial:
			// radial controls likewise just need the entire frame:

			setMiniControlBounds(cellframe, identifier: kRadialStartControlID)
			setMiniControlBounds(cellframe, identifier: kRadialStartControlID)

		case .sweep:
			// sweep controls:

			setMiniControlBounds(cframe, identifier: kSweepAngleControlID)
			setMiniControlBounds(cellframe, identifier: kSweepCentreControlID)

			// only the segment slider has a complex bounds calc:

			var sr = cframe.insetBy(dx: 40, dy: 50)
			sr.size.height = 12
			sr.origin.y = cframe.origin.y + (cframe.size.height * 0.63)
			
			setMiniControlBounds(sr, identifier: kSweepSegmentsControlID)
			
		default:
			break
		}
	}
	
	/// Sets the mini control with `key` identifier to have the bounds `br`
	func setMiniControlBounds(_ br: NSRect, identifier key: String) {
		if let mc = miniControls?.control(forKey: key) {
			mc.bounds = br
		}
	}
	
	/// given the mode which is set by the owning GCGradientWell, this draws the controls in the appropriate cluster.
	func drawMiniControls(for mode: GCSGradientWell.Mode) {
		controlCluster(for: mode)?.draw()
	}
	
	func controlCluster(for mode: GCSGradientWell.Mode) -> GCMiniControlCluster? {
		switch mode {
		case .angle:
			return miniControls?.control(forKey: kLinearControlsClusterID) as? GCMiniControlCluster
			
		case .radial:
			return miniControls?.control(forKey: kRadialControlsClusterID) as? GCMiniControlCluster

		case .sweep:
			return miniControls?.control(forKey: kSweepControlsClusterID) as? GCMiniControlCluster

		default:
			return nil
		}
	}
	
	func miniControl(forIdentifier key: String) -> GCMiniControl? {
		return miniControls?.control(forKey: key)
	}
	
	/// sets the minicontrol values in `mode` cluster to match the current gradient
	func updateMiniControls(for mode: GCSGradientWell.Mode) {
		updatingControls = true
		
		switch mode {
		case .angle:
			if let gradAngle = gradient?.angle {
				miniControl(forIdentifier: kLinearAngleControlID)?.value = gradAngle
			}
			
		case .radial:
			LogEvent(.stateEvent, "setting up radial controls")
			guard var rc = miniControls?.control(forKey: kRadialStartControlID) as? GCMiniRadialControl2,
			let gradient = self.gradient else {
				break
			}
			
			rc.setRingRadiusScale(0.85)
			rc.centre = gradient.mapPoint(gradient.radialStartingPoint, to: controlBoundsRect)
			rc.radius = gradient.radialStartingRadius * controlBoundsRect.size.width
			rc.tabColor = gradient.color(atValue: 0)
			
			rc = miniControls!.control(forKey: kRadialEndControlID) as! GCMiniRadialControl2
			
			rc.centre = gradient.mapPoint(gradient.radialEndingPoint, to: controlBoundsRect)
			rc.radius = gradient.radialEndingRadius * controlBoundsRect.size.width
			rc.tabColor = gradient.color(atValue: 1)
			
		case .sweep:
			// sweep controls:

			guard let rc = miniControls?.control(forKey: kSweepCentreControlID) as? GCMiniRadialControls,
			let gradient = self.gradient else {
				break
			}
			rc.centre = gradient.mapPoint(gradient.radialStartingPoint, to: controlBoundsRect)
			
			let seg = 100 //[[self gradient] numberOfAngularSegments];
			var v = CGFloat(seg) / 50.0
			
			if seg < 4 {
				v = 0
			}
			
			miniControl(forIdentifier: kSweepSegmentsControlID)?.value = v
			miniControl(forIdentifier: kSweepAngleControlID)?.value = gradient.angle
			
		default:
			break
		}
		
		updatingControls = false
	}
	
	// MARK: -
	func proxyIconRect(inCellFrame rect:NSRect) -> NSRect {
		let br = rect.insetBy(dx: 8, dy: 9)
		let ficon = NSImage(named: NSImage.Name(rawValue: "fileiconsmall"))
		var ir = NSRect()

		if let ficon = ficon {
			ir.size = ficon.size
		}
		ir.origin.x = br.maxX - ir.size.width
		ir.origin.y = br.maxY - ir.size.height
		
		return ir
	}
	
	// MARK: -
	@objc func setControlVisible(_ vis: Bool) {
		miniControls?.visible = vis
		controlView?.needsDisplay = true
	}
	
	// MARK: - As an DKSGradientCell

	override func drawInterior(withFrame cellFrame: NSRect, in controlView: NSView) {
		if gradient != nil {
			super.drawInterior(withFrame: cellFrame, in: controlView)
			
			controlBoundsRect = cellFrame
			
			if let control = self.controlView as? GCSGradientWell {
				control.setupTrackingRect()
			
				//[self setMiniControlBoundsWithCellFrame:cellFrame forMode:[control controlMode]]
				updateMiniControls(for: control.controlMode)
				
				NSBezierPath.clip(cellFrame.insetBy(dx: 8, dy: 8))
				drawMiniControls(for: control.controlMode)
				
				// if proxy icon flag set, draw it

				if control.displaysProxyIcon,
					let ficon = NSImage(named: NSImage.Name(rawValue: "fileiconsmall")) {
					ficon.draw(in: proxyIconRect(inCellFrame: cellFrame), from: .zero, operation: .sourceAtop, fraction: 0.8)
				}
			}
		}
	}
	
	// MARK: - As an NSCell
	
	override func continueTracking(last lastPoint: NSPoint, current currentPoint: NSPoint, in controlView: NSView) -> Bool {
		if let cv = controlView as? GCSGradientWell,
			hitPart == .miniControl {
			let cmode = cv.controlMode
			let cc = controlCluster(for: cmode)
			
			return cc?.mouseDragged(at: currentPoint, inPart: GCControlHitTest(0), modifierFlags: NSEvent.ModifierFlags(rawValue: UInt(mouseDownFlags))) ?? false
		}
		return false
	}
	
	override var mouseDownFlags: Int {
		return Int(sMFlags)
	}
	
	override func startTracking(at startPoint: NSPoint, in controlView: NSView) -> Bool {
		LogEvent(.reactiveEvent, "cell starting tracking...");
		
		hitPart = .other
		
		guard let cv = controlView as? GCSGradientWell else {
			return true
		}
		// hit in proxy icon?

		if cv.displaysProxyIcon {
			let ir = proxyIconRect(inCellFrame: controlBoundsRect)
			
			if ir.contains(startPoint) {
				hitPart = .proxyIcon
				return true
			}
		}
		
		let cmode = cv.controlMode
		if let cc = controlCluster(for: cmode), cc.mouseDown(at: startPoint, inPart: 0, modifierFlags: NSEvent.ModifierFlags(rawValue: UInt(mouseDownFlags))) {
			hitPart = .miniControl // for any mini-control
		}
		return true
	}
	
	override func stopTracking(last lastPoint: NSPoint, current stopPoint: NSPoint, in controlView: NSView, mouseIsUp flag: Bool) {
		if let cv = controlView as? GCSGradientWell, hitPart == .miniControl {
			let cmode = cv.controlMode
			let cc = controlCluster(for: cmode)
			
			cc?.mouseUp(at: stopPoint, inPart: 0, modifierFlags: NSEvent.ModifierFlags(rawValue: UInt(mouseDownFlags)))
		}
	}
	
	override func trackMouse(with theEvent: NSEvent, in cellFrame: NSRect, of controlView: NSView, untilMouseUp flag: Bool) -> Bool {
		let p = controlView.convert(theEvent.locationInWindow, from: nil)
		sMFlags = theEvent.modifierFlags.rawValue
		miniControls?.flagsChanged(NSEvent.ModifierFlags(rawValue: sMFlags))
		
		if startTracking(at: p, in: controlView) {
			var event: NSEvent?
			var loop = true
			var currentPoint: NSPoint
			
			enableCache = false
			let mask: NSEvent.EventTypeMask = [.leftMouseUp, .leftMouseDragged, .flagsChanged]
			var lastPoint = p
			
			while loop {
				event = controlView.window?.nextEvent(matching: mask)
				guard let event2 = event else {
					continue
				}
				
				currentPoint = controlView.convert(event2.locationInWindow, from: nil)
				sMFlags = event2.modifierFlags.rawValue
				
				switch event2.type {
				case .leftMouseUp:
					stopTracking(last: lastPoint, current: currentPoint, in: controlView, mouseIsUp: true)
					loop = false
					
				case .leftMouseDragged:
					loop = continueTracking(last: lastPoint, current: currentPoint, in: controlView)

					if !loop {
						stopTracking(last: lastPoint, current: currentPoint, in: controlView, mouseIsUp: false)
						if hitPart == .other {
							miniControls?.visible = false
							(controlView as? DKSGradientDragging)?.initiateGradientDrag(with: theEvent)
						}
					}
					
				case .flagsChanged:
					miniControls?.flagsChanged(event2.modifierFlags)
					
				default:
					break
				}
				
				lastPoint = currentPoint
			}
			
			controlView.window?.discardEvents(matching: mask, before: event)
			hitPart = .none
			enableCache = true
			miniControls?.flagsChanged([])
		}
		
		LogEvent(.reactiveEvent, "cell ended tracking");
		
		return true
	}
	
	// MARK: - As an NSObject
	
	override init() {
		super.init()
		setupMiniControls()
	}
	
	required init(coder: NSCoder) {
		super.init(coder: coder)
		setupMiniControls()
	}
	
	
	// MARK: - As a GCMiniControl delegate
	
	/// delegate method called for a change in any mini-control value. route the result to the appropriate
	/// setting. Note - no need to call for redisplay, that has been done.
	func miniControl(_ mc: GCMiniControl!, didChangeValue newValue: Any!) {
		if !updatingControls {
			LogEvent(.infoEvent, "miniControl ‘\(mc.identifier)’ didChangeValue ‘\(newValue)’")
			
			setControlledAttribute(from: mc)

			if let control = self.controlView as? GCSGradientWell {
				control.syncGradientToControlSettings()
			}
		}
	}
	
	func miniControlWillUpdateInfoWindow(_ mc: GCMiniControl!, withValue val: CGFloat) -> CGFloat {
		if mc.identifier == kSweepSegmentsControlID {
			var seg = mc.value * 50
			if seg < 4 {
				seg = 0
			}
			
			return seg
		} else {
			return val
		}
	}
}
